### PR TITLE
Parallel section sampling

### PR DESCRIPTION
**Pull request by MalteMederacke**
This parallelises the sampling from the graph edges and allows for lazy loading of the relevant image section.

Leads to:
- Significant increase in performance
- Out of memory sampling from large volumes

### CODE REVIEW

File: src/skeleplex/graph/sample.py [hunk 1]
```diff
@@ -109,3 +109,72 @@ def sample_volume_at_coordinates(
...
+        Array of shape (*grid_shape)
+    """
+    batch, *grid_shape, _ = coordinates.shape
+
+    min_coords = np.floor(coordinates.min(axis=(0, 1, 2))).astype(int)
+    max_coords = np.ceil(coordinates.max(axis=(0, 1, 2))).astype(int)
+
+    # Clip to valid range
+    min_coords = np.clip(min_coords, 0, np.array(volume.shape) - 1)
+    max_coords = np.clip(max_coords, 1, np.array(volume.shape))  # avoid empty slices
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/sample.py:154]: Does this prevent the fill value from being used for coordinates outside of the array indices?

File: src/skeleplex/graph/skeleton_graph.py [hunk 2]
```diff
@@ -209,6 +229,52 @@ def orient_splines(graph: nx.DiGraph) -> nx.DiGraph:
...
+    volume_path,
+    segmentation_path,
+    positions,
+    slice_size,
+    pixel_size,
+    interpolation_order,
+    image_voxel_size_um,
+    approx,
+):
+    """Helper function for parallel execution."""
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py]: Please complete the docstring with parameters, type annotations, and returns.

File: src/skeleplex/graph/skeleton_graph.py [hunk 3]
```diff
@@ -209,6 +229,52 @@ def orient_splines(graph: nx.DiGraph) -> nx.DiGraph:
...
+            positions,
+            grid_shape=(slice_size, slice_size),
+            grid_spacing=(pixel_size, pixel_size),
+            sample_interpolation_order=0,
+            image_voxel_size_um=image_voxel_size_um,
+            sample_fill_value=0,
+            approx=approx,
+        )
+
+    return (u, v), image_slice, segmentation_slice
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py:308]: I think it would be cleaner to just have the function accept a single path and then when you need to do the image and segmentation (and other images) just call the function multiple times.

File: src/skeleplex/graph/skeleton_graph.py [hunk 4]
```diff
@@ -453,6 +545,9 @@ def sample_volume_slices_from_spline(
         segmentation : np.ndarray | None
             The segmentation to sample slices from.
             If None, only the image is sampled.
+        approx : bool
+            If True, use the approximate spline evaluation.
+            If False, use the exact spline evaluation.
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py]: Please add a note explaining the approximation.

File: src/skeleplex/graph/skeleton_graph.py [hunk 5]
```diff
@@ -509,10 +605,122 @@ def sample_volume_slices_from_spline(
...
+            )
+
+        # Use Dask delayed to process edges in parallel
+        tasks = [delayed(process_edge)(edge) for edge in edges_to_process]
+
+        # Compute all tasks in parallel
+        logger.info(f"Processing {len(tasks)} edges in parallel.")
+        with ProgressBar():
+            results = dask.compute(
+                *tasks, num_workers=os.cpu_count() - 2, scheduler="processes"
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py]: I think the num_workers should be exposed as a parameter

File: src/skeleplex/graph/skeleton_graph.py [hunk 6]
```diff
@@ -509,10 +605,122 @@ def sample_volume_slices_from_spline(
...
+        image_voxel_size_um = self.voxel_size_um
+        origin = self.origin
+        if not origin:
+            raise ValueError("No origin node provided. Please set origin.")
+
+        if not image_voxel_size_um:
+            logger.warning("No voxel size provided. Assuming pixel size is 1 µm.")
+            image_voxel_size_um = (1, 1, 1)
+
+        graph = self.graph.copy()
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py]: Why is it necessary to copy the graph?

File: src/skeleplex/graph/skeleton_graph.py [hunk 7]
```diff
@@ -509,10 +605,122 @@ def sample_volume_slices_from_spline(
...
+                    approx=approx,
                 )
                 segmentation_slice_dict[(u, v)] = segmentation_slice
 
         if segmentation is not None:
             return image_slice_dict, segmentation_slice_dict
         else:
             return image_slice_dict
+
+    def sample_volume_slices_from_spline_parallel(
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py:655]: Please add a test. It can be simple (e.g., a line along the z axis with z slices of different constant values). I would be sure to use a pixel size that isn't 1 to make sure the conversions are being done correctly. You may need to set workers to 1 to make it work on CI.

File: src/skeleplex/measurements/branches.py [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+import h5py
+import numpy as np
+import skimage as ski
+import torch
+from tqdm import tqdm
+import concurrent.futures
+
+
+from sam2.build_sam import build_sam2
+from sam2.sam2_image_predictor import SAM2ImagePredictor
```
Comment (kevinyamauchi) [inline src/skeleplex/measurements/branches.py:13]: We should add sam2 as an optional dependency. Maybe under the torch group?

File: src/skeleplex/measurements/branches.py [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+        image_slice_filt = np.delete(image_slices, index_to_remove, axis=0)
+        label_slices_filt = np.delete(label_slices_filt, index_to_remove, axis=0)
+
+        with h5py.File(os.path.join(save_path, file), "w") as f:
+            f.create_dataset("image", data=image_slice_filt)
+            f.create_dataset("segmentation", data=label_slices_filt)
+
+
+def filter_for_iterative_lumens(data_path, save_path):
+    """Filter for iterative lumens across multiple files in parallel."""
```
Comment (kevinyamauchi) [inline src/skeleplex/measurements/branches.py]: please complete the docstring

File: src/skeleplex/measurements/branches.py [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+
+    # Pack arguments for parallel processing
+    file_args = [(f, data_path, save_path) for f in files]
+
+    with concurrent.futures.ProcessPoolExecutor(max_workers=12) as executor:
+        list(executor.map(filter_file_for_iterative_lumens, file_args))
+
+
+def filter_file_for_iterative_lumens(args):
+    """Filter a single HDF5 file for iterative lumens."""
```
Comment (kevinyamauchi) [inline src/skeleplex/measurements/branches.py]: please complete the docstring

File: src/skeleplex/measurements/branches.py [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+    data_path,
+    save_path,
+    sam_checkpoint_path,
+    resnet_predictor,
+    eccentricity_thresh=0.7,
+    circularity_thresh=0.5,
+    find_lumen=True,
+):
+    """
+    Filter and segment the lumen in the image slices.
```
Comment (kevinyamauchi) [inline src/skeleplex/measurements/branches.py:32]: Please explain what you mean by lumen and what the filtering is

File: src/skeleplex/measurements/branches.py [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+    )
+
+
+def add_measurements_from_h5_to_graph(graph_path, input_path):
+    """
+    Add measurements from HDF5 files to the skeleton graph.
+
+    The slice names need to be in the format:
+
+    {base}_{name}_{start_node}_{end_node}.h5
```
Comment (kevinyamauchi) [inline src/skeleplex/measurements/branches.py:440]: Please briefly explain where the h5 files should come from (i.e., which functions create valid files).

File: src/skeleplex/graph/sample.py [hunk 13]
```diff
@@ -109,3 +109,72 @@ def sample_volume_at_coordinates(
     sampled_volume = sampled_volume.reshape(*grid_shape, batch)
     # and retranspose to get batch back to the 0th dimension
     return einops.rearrange(sampled_volume, "... batch -> batch ...")
+
+
+def sample_volume_at_coordinates_lazy(
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/sample.py:114]: I think it would be nice to have a test for this. Just something simple where the box goes outside of the image to verify that the fill value is being implemented correctly. You can use interpolation_order = 1 to make it easier.

**Review by kevinyamauchi — APPROVED** [review]
Thanks! I left some comments! Once you address them, please feel free to merge.

File: src/skeleplex/graph/skeleton_graph.py (hunk 3 above)
Comment (MalteMederacke) [reply src/skeleplex/graph/skeleton_graph.py:308]: I set it up in a way, such the segmentation and branch section is then stored together in an h5 container is always neatly together. This reduces the number of files by half to, so I actually like it a lot how it is. And if one doesn't want to sample a segmentation or is happy to only do the segmentation it still works, so what would be the downside in keeping it this way?

File: src/skeleplex/measurements/branches.py (hunk 8 above)
Comment (MalteMederacke) [reply src/skeleplex/measurements/branches.py:13]: I added a optional dependency in the other PR... Moved it over here, I added it to an extra group, because SAM is not on pip and requires manual installation of the weights, so should be a conciouse process.

File: src/skeleplex/graph/sample.py (hunk 13 above)
Comment (MalteMederacke) [reply src/skeleplex/graph/sample.py:114]: Added a test

File: src/skeleplex/graph/skeleton_graph.py (hunk 7 above)
Comment (MalteMederacke) [reply src/skeleplex/graph/skeleton_graph.py:655]: added a test for the normal sampling and the spline sampling

**Review by MalteMederacke — COMMENTED** [review]
> Does this prevent the fill value from being used for coordinates outside of the array indices? 

No, this just avoids loading nothing from the volume I think. fill value behaves as expected, the test should cover that, too.

**Review by MalteMederacke — COMMENTED** [review]
> Does this prevent the fill value from being used for coordinates outside of the array indices? 

No, this just avoids loading nothing from the volume I think. fill value behaves as expected, the test should cover that, too.

**Comment by kevinyamauchi** [comment]
Thanks for the updates! Merging now.
